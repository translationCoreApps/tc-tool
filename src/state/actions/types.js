export const LOCALE_LOADED = 'LOCALE_LOADED';
