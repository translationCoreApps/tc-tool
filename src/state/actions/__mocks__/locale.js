import {initialize, addTranslationForLanguage} from 'react-localize-redux';
import * as types from '../types';

export const setLocaleLoaded = () => ({
    type: types.LOCALE_LOADED
});

export const loadLocalization = jest.fn(() => {
    return dispatch => {
        const languages = ['en_US'];
        dispatch(initialize(languages));
        dispatch(addTranslationForLanguage({ hello: 'hello'}, 'en_US'));
        dispatch(setLocaleLoaded());
    };
});