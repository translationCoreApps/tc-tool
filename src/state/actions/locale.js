import fs from 'fs-extra';
import path from 'path';
import {
  addTranslationForLanguage,
  initialize,
  setActiveLanguage
} from 'react-localize-redux';
import osLocale from 'os-locale';
import _ from 'lodash';
import * as types from './types';
import {getLanguages, getTranslations} from '../reducers';
import {batchActions} from 'redux-batched-actions';

const DEFAULT_LOCALE = 'en_US';

/**
 * The handler for missing translations.
 * @param key
 * @param languageCode
 */
const onMissingTranslation = (key, languageCode) => {
  console.error(`Tool missing locale key "${key}" for language ${languageCode}`,
    new Error().stack);
};

/**
 * Splits a locale filename into it's identifiable pieces
 * @param {string} fileName the locale file name (basename)
 * @return {{langName, langCode, shortLangCode}}
 */
const explodeLocaleName = (fileName) => {
  let title = fileName.replace(/\.json/, '');
  let langName = title.split('-')[0];
  let langCode = title.split('-')[1];
  let shortLangCode = langCode.split('_')[0];
  return {langName, langCode, shortLangCode};
};

/**
 * Injects additional information into the translation
 * that should not otherwise be translated. e.g. legal entities
 * @param {object} translation localized strings
 * @param {string} fileName the name of the locale file including the file extension.
 * @param {list} nonTranslatableStrings a list of non-translatable strings to inject
 * @return {object} the enhanced translation
 */
const enhanceTranslation = (
  translation, fileName, nonTranslatableStrings = []) => {
  const {langName, langCode, shortLangCode} = explodeLocaleName(fileName);
  return {
    ...translation,
    '_': {
      'language_name': langName,
      'short_locale': shortLangCode,
      'locale': langCode,
      ...nonTranslatableStrings
    }
  };
};

/**
 * Indicates the locale has been completely loaded
 * @return {{type: string}}
 */
export const setLocaleLoaded = () => ({
  type: types.LOCALE_LOADED
});

/**
 * This thunk loads the localization data
 * and initializes the localization library.
 *
 * The default language is english.
 * TODO: for now we are loading all translations up-front. However we could instead load one at a time as needed in `setLanguage` for better performance.
 *
 * @param {string} localeDir directory containing locale files
 * @param {string} appLanguage the language code that will be enabled by default
 * @return {function(*)}
 */
export const loadLocalization = (localeDir, appLanguage = null) => {
  return (dispatch) => {
    if (!fs.existsSync(localeDir)) {
      throw new Error(`Tool missing locale dir at ${localeDir}`);
    }
    const nonTranslatableFile = 'nonTranslatable.json';
    const items = fs.readdirSync(localeDir);
    // load locale
    let languages = [];
    let translations = {};
    if (!items) {
      throw new Error(`Tool found no localization files in ${localeDir}`);
    }
    for (let file of items) {
      if (!file.endsWith('.json')) {
        // don't warn if readme or NonTranslatable.js
        if (!file.endsWith('.md') && !file.endsWith('.js')) {
          console.warn(`Tool skipping invalid localization file ${file}`);
        }
        continue;
      }
      if (file === nonTranslatableFile) { // skip the non-translatable file
        continue;
      }
      const localeFile = path.join(localeDir, file);
      try {
        let translation = JSON.parse(fs.readFileSync(localeFile));

        const nonTranslatablePath = path.join(localeDir, nonTranslatableFile);
        let nonTranslatable = {};
        if (fs.existsSync(nonTranslatablePath)) {
          nonTranslatable = JSON.parse(fs.readFileSync(nonTranslatablePath));
        }

        translation = enhanceTranslation(translation, file, nonTranslatable);

        const {langCode, shortLangCode} = explodeLocaleName(file);
        languages.push(langCode);
        translations[langCode] = translation;

        // include short language names for wider locale compatibility
        if (_.indexOf(languages, shortLangCode) === -1) {
          languages.push(shortLangCode);
          translations[shortLangCode] = translation;
        }
      } catch (e) {
        console.warn(`Tool failed to load localization ${localeFile}: ${e}`);
      }
    }

    // init locale
    const namedLanguages = languages.map((code) => {
      return {
        code,
        name: translations[code]['_']['language_name']
      };
    });
    dispatch(initialize(namedLanguages, {
      defaultLanguage: DEFAULT_LOCALE,
      missingTranslationCallback: onMissingTranslation
    }));
    const addTranslationActions = [];
    for (const languageCode in translations) {
      if (translations.hasOwnProperty(languageCode)) {
        addTranslationActions.push(addTranslationForLanguage(translations[languageCode], languageCode));
      }
    }
    dispatch(batchActions(addTranslationActions));
    if (appLanguage !== DEFAULT_LOCALE) {
      setActiveLocale(dispatch, appLanguage, languages, translations);
    }
    dispatch(setLocaleLoaded());
  };
};

/**
 * Safely sets the active language by falling back to an equivalent locale if
 * needed.
 *
 * @param dispatch
 * @param {string} locale the locale to set
 * @param {list} languages a list of loaded languages
 * @param {object} translations a dictionary of loaded translations
 * @return {bool} returns true of the language was successfully set.
 */
const setActiveLanguageSafely = (dispatch, locale, languages, translations) => {
  const shortLocale = locale.split('_')[0];
  if (_.indexOf(languages, locale) >= 0) {
    // matched locale
    dispatch(setActiveLanguage(locale));
  } else if (_.indexOf(languages, shortLocale) >= 0) {
    // equivalent locale
    let equivalentLocale = translations[shortLocale]['_']['locale'];
    console.info(`Using equivalent locale: ${equivalentLocale}`);
    dispatch(setActiveLanguage(equivalentLocale));
  } else {
    console.error(`Tool found no translations for locale: ${locale}`);
    return false;
  }
  return true;
};

/**
 * Sets the active locale.
 * This will fallback to the system os, then english if the locale is not found.
 *
 * @param dispatch
 * @param {string} locale - the language code to set. This can be null
 * @param languages
 * @param translations
 * @return {function(*=, *)}
 */
const setActiveLocale = (dispatch, locale, languages, translations) => {
  const systemLocale = osLocale.sync();
  const locales = [locale, systemLocale, 'en_US'];
  let foundLocale = false;
  for (const langCode of locales) {
    // TRICKY: make sure the input locale was not null
    if (langCode &&
      setActiveLanguageSafely(dispatch, langCode, languages, translations)) {
      if (langCode !== locale) {
        console.warn(
          `Tool could not find locale ${locale}. Falling back to ${langCode}`);
      }
      foundLocale = true;
      break;
    }
  }
  if (!foundLocale) {
    console.error('Tool was unable to find suitable locale.');
  }
};

/**
 * Sets the active locale.
 * This will fallback to the system os, then english if the locale is not found.
 *
 * @param {string} locale - the language code to set
 * @return {function(*=, *)}
 */
exports.setActiveLocale = (locale) => {
  return (dispatch, getState) => {
    const state = getState();
    const translations = getTranslations(state);
    const languages = _.map(getLanguages(state), 'code');
    setActiveLocale(dispatch, locale, languages, translations);
  };
};
