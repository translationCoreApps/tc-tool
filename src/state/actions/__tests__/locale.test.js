jest.unmock('react-localize-redux');
import * as actions from '../locale';
import * as types from '../types';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import path from 'path';
import _ from 'lodash';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('actions', () => {

  it('should create an action to set the locale loaded', () => {
    const expectedAction = {
      type: types.LOCALE_LOADED
    };
    expect(actions.setLocaleLoaded()).toEqual(expectedAction);
  });

  describe('create an action to initialize the locale', () => {

    it('should inject non-translatable strings', () => {
      global.console = {warn: jest.fn()};
      let localeDir = path.join(__dirname, './locale/');
      const store = mockStore({});
      store.dispatch(actions.loadLocalization(localeDir, 'en_US'));

      let addTranslationActions = store.getActions().map(action => {
        if (action.type ===
          '@@localize/ADD_TRANSLATION_FOR_LANGUGE') return action;
      });
      addTranslationActions = _.compact(addTranslationActions);
      expect(addTranslationActions).toHaveLength(4);
      const action = addTranslationActions[0];
      const translation = action.payload.translation;
      expect(translation).toHaveProperty('_');
      expect(translation._).toHaveProperty('locale');
      expect(translation._).toHaveProperty('language_name');
      expect(console.warn).toBeCalled();
    });

    it('should not use the system locale', () => {
      global.console = {warn: jest.fn()};
      let defaultLanguage = 'en_US';
      let localeDir = path.join(__dirname, './locale/');
      const expectedActionTypes = [
        '@@localize/INITIALIZE',
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', //en_US
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', // for short locale addition
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', //na_NA
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', // for short locale addition
        '@@tool/LOCALE_LOADED'
      ];
      const store = mockStore({});
      store.dispatch(actions.loadLocalization(localeDir, defaultLanguage));
      let receivedActionTypes = store.getActions().map(action => {
        return action.type;
      });
      expect(receivedActionTypes).toEqual(expectedActionTypes);
      expect(console.warn).toBeCalled();
    });

    it('should use the system locale', () => {
      global.console = {warn: jest.fn()};
      let localeDir = path.join(__dirname, './locale/');
      const expectedActionTypes = [
        '@@localize/INITIALIZE',
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', //en_US
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', // for short locale addition
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', //na_NA
        '@@localize/ADD_TRANSLATION_FOR_LANGUGE', // for short locale addition
        '@@localize/SET_ACTIVE_LANGUAGE',
        '@@tool/LOCALE_LOADED'
      ];
      const store = mockStore({});
      store.dispatch(actions.loadLocalization(localeDir));
      let receivedActionTypes = store.getActions().map(action => {
        return action.type;
      });
      expect(receivedActionTypes).toEqual(expectedActionTypes);
      expect(console.warn).toBeCalled();
    });

    it('should reject if locale dir is missing', () => {
      let localeDir = null;
      const store = mockStore({});
      expect(() => store.dispatch(actions.loadLocalization(localeDir))).toThrowError('Tool missing locale dir at null');
    });

    it('should use an equivalent locale', () => {
      global.console = {info: jest.fn(), warn: jest.fn()};
      let defaultLanguage = 'na_MISSING';
      let localeDir = path.join(__dirname, './locale/');
      const expectedActionTypes = [
        {type: '@@localize/INITIALIZE', languageCode: undefined},
        {
          type: '@@localize/ADD_TRANSLATION_FOR_LANGUGE',
          languageCode: undefined
        }, // en_US
        {
          type: '@@localize/ADD_TRANSLATION_FOR_LANGUGE',
          languageCode: undefined
        }, // for short locale addition
        {
          type: '@@localize/ADD_TRANSLATION_FOR_LANGUGE',
          languageCode: undefined
        }, // na_NA
        {
          type: '@@localize/ADD_TRANSLATION_FOR_LANGUGE',
          languageCode: undefined
        }, // for short locale addition
        {type: '@@localize/SET_ACTIVE_LANGUAGE', languageCode: 'na_NA'},
        {type: '@@tool/LOCALE_LOADED'}
      ];
      const store = mockStore({});
      store.dispatch(actions.loadLocalization(localeDir, defaultLanguage));
      let receivedActionTypes = store.getActions().map(action => {
        if (action.type.startsWith('@@localize')) {
          return {type: action.type, languageCode: action.payload.languageCode};
        } else {
          return {type: action.type};
        }
      });
      expect(receivedActionTypes).toEqual(expectedActionTypes);
      expect(console.info).toBeCalled();
      expect(console.warn).toBeCalled();
    });
  });
});
