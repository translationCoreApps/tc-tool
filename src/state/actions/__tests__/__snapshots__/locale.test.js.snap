// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions create an action to initialize the locale should inject non-translatable strings 1`] = `
Array [
  Object {
    "payload": Object {
      "languages": Array [
        Object {
          "code": "en_US",
          "name": "English",
        },
        Object {
          "code": "en",
          "name": "English",
        },
        Object {
          "code": "na_NA",
          "name": "Name",
        },
        Object {
          "code": "na",
          "name": "Name",
        },
      ],
      "options": Object {
        "defaultLanguage": "en_US",
        "missingTranslationCallback": [Function],
      },
    },
    "type": "@@localize/INITIALIZE",
  },
  Object {
    "meta": Object {
      "batch": true,
    },
    "payload": Array [
      Object {
        "payload": Object {
          "language": "en_US",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "en",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na_NA",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
    ],
    "type": "BATCHING_REDUCER.BATCH",
  },
  Object {
    "type": "@@tool/LOCALE_LOADED",
  },
]
`;

exports[`actions create an action to initialize the locale should not use the system locale 1`] = `
Array [
  Object {
    "payload": Object {
      "languages": Array [
        Object {
          "code": "en_US",
          "name": "English",
        },
        Object {
          "code": "en",
          "name": "English",
        },
        Object {
          "code": "na_NA",
          "name": "Name",
        },
        Object {
          "code": "na",
          "name": "Name",
        },
      ],
      "options": Object {
        "defaultLanguage": "en_US",
        "missingTranslationCallback": [Function],
      },
    },
    "type": "@@localize/INITIALIZE",
  },
  Object {
    "meta": Object {
      "batch": true,
    },
    "payload": Array [
      Object {
        "payload": Object {
          "language": "en_US",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "en",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na_NA",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
    ],
    "type": "BATCHING_REDUCER.BATCH",
  },
  Object {
    "type": "@@tool/LOCALE_LOADED",
  },
]
`;

exports[`actions create an action to initialize the locale should use an equivalent locale 1`] = `
Array [
  Object {
    "payload": Object {
      "languages": Array [
        Object {
          "code": "en_US",
          "name": "English",
        },
        Object {
          "code": "en",
          "name": "English",
        },
        Object {
          "code": "na_NA",
          "name": "Name",
        },
        Object {
          "code": "na",
          "name": "Name",
        },
      ],
      "options": Object {
        "defaultLanguage": "en_US",
        "missingTranslationCallback": [Function],
      },
    },
    "type": "@@localize/INITIALIZE",
  },
  Object {
    "meta": Object {
      "batch": true,
    },
    "payload": Array [
      Object {
        "payload": Object {
          "language": "en_US",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "en",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na_NA",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
    ],
    "type": "BATCHING_REDUCER.BATCH",
  },
  Object {
    "payload": Object {
      "languageCode": "na_NA",
    },
    "type": "@@localize/SET_ACTIVE_LANGUAGE",
  },
  Object {
    "type": "@@tool/LOCALE_LOADED",
  },
]
`;

exports[`actions create an action to initialize the locale should use the system locale 1`] = `
Array [
  Object {
    "payload": Object {
      "languages": Array [
        Object {
          "code": "en_US",
          "name": "English",
        },
        Object {
          "code": "en",
          "name": "English",
        },
        Object {
          "code": "na_NA",
          "name": "Name",
        },
        Object {
          "code": "na",
          "name": "Name",
        },
      ],
      "options": Object {
        "defaultLanguage": "en_US",
        "missingTranslationCallback": [Function],
      },
    },
    "type": "@@localize/INITIALIZE",
  },
  Object {
    "meta": Object {
      "batch": true,
    },
    "payload": Array [
      Object {
        "payload": Object {
          "language": "en_US",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "en",
          "translation": Object {
            "_": Object {
              "language_name": "English",
              "locale": "en_US",
              "short_locale": "en",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na_NA",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
      Object {
        "payload": Object {
          "language": "na",
          "translation": Object {
            "_": Object {
              "language_name": "Name",
              "locale": "na_NA",
              "short_locale": "na",
            },
          },
        },
        "type": "@@localize/ADD_TRANSLATION_FOR_LANGUGE",
      },
    ],
    "type": "BATCHING_REDUCER.BATCH",
  },
  Object {
    "payload": Object {
      "languageCode": "en_US",
    },
    "type": "@@localize/SET_ACTIVE_LANGUAGE",
  },
  Object {
    "type": "@@tool/LOCALE_LOADED",
  },
]
`;
